backend/src/apiserver/resource/resource_manager.go

```diff
@@ -1152,3 +1154,161 @@ func (r *ResourceManager) getNamespaceFromExperiment(references []*api.ResourceR
+							MountPath: "/var/lib/docker/containers",
+							ReadOnly:  true,
+						},
+						{
+							Name:      "varlibkubeletpods",
+							MountPath: "/var/lib/kubelet/pods",
+							ReadOnly:  true,
+						},
+						{
+							Name:      "varlogpods",
+							MountPath: "/var/log/pods",
+							ReadOnly:  true,
+						},
+					}
+					task.TaskSpec.StepTemplate.VolumeMounts = append(task.TaskSpec.StepTemplate.VolumeMounts, loggingVolumeMounts...)
+				}
+
+				// Upload Artifacts if the artifact is enabled and the annoations are present
+				if hasArtifacts && len(artifacts) > 0 && trackArtifacts {
+					for _, artifact := range artifacts {
+						if len(artifact) == 2 {
+							artifactScript += fmt.Sprintf("push_artifact %s %s\n", artifact[0], artifact[1])
+						}
+					}
+				}
+
+				// Strip EOF if enabled, do it after artifact upload since it only applies to parameter outputs
+				if hasArtifacts && len(artifacts) > 0 && stripEOF {
+					for _, artifact := range artifacts {
+						if len(artifact) == 2 {
+							// TODO: Add EOF newline stripping logics. we need a safe and simple command that can run on the
+							//       minio/mc image. Here we need 1. Check is EOF is newline, 2. Remove EOF newlines and
+							//       update the file. 3. Done with one line bash command. 4. File path is a Tekton substitution
+							//       variable such as $(results.project.path) before runtime.
+							//       We may need to run another image that has python or perl to include these logics.
+							//       This feature is better to implement in Tekton controller since KFP can't inject logics
+							//       during pipeline runtime.
+							//       The below solution is in experimental stage and didn't cover all edge cases.
+							artifactScript += fmt.Sprintf("strip_eof %s %s\n", artifact[0], artifact[1])
+						}
+					}
+				}
+
+				// Define post-processing step
+				step := workflowapi.Step{Container: corev1.Container{
+					Name:  "copy-artifacts",
+					Image: "minio/mc",
+					Env: []corev1.EnvVar{
+						r.getObjectFieldSelector("ARTIFACT_BUCKET", "metadata.annotations['tekton.dev/artifact_bucket']"),
+						r.getObjectFieldSelector("ARTIFACT_ENDPOINT", "metadata.annotations['tekton.dev/artifact_endpoint']"),
+						r.getObjectFieldSelector("ARTIFACT_ENDPOINT_SCHEME", "metadata.annotations['tekton.dev/artifact_endpoint_scheme']"),
+						r.getObjectFieldSelector("PIPELINETASK", "metadata.labels['tekton.dev/pipelineTask']"),
+						r.getObjectFieldSelector("PIPELINERUN", "metadata.labels['tekton.dev/pipelineRun']"),
+						r.getObjectFieldSelector("PODNAME", "metadata.name"),
+						r.getSecretKeySelector("AWS_ACCESS_KEY_ID", "mlpipeline-minio-artifact", "accesskey"),
+						r.getSecretKeySelector("AWS_SECRET_ACCESS_KEY", "mlpipeline-minio-artifact", "secretkey"),
+					},
+				},
+					Script: artifactScript,
+				}
+				task.TaskSpec.Steps = append(task.TaskSpec.Steps, step)
+			}
+		}
+	}
+	return nil
+}
```

Comment:
120 lines and 6 levels of indentation is quite a lot. Many fragments are very "local" too (e.g. creating the post-processing step, injecting mounts to step template etc). Looks like a good candidate for refactoring.

Reply:
Thanks, I rewrote this function as part of making the "minio/mc" image customizable.